src/k2/persistence/logStream/LogStream.cpp

```diff
@@ -0,0 +1,343 @@
+        for (auto& response: responses){
+            auto& [status, plogId] = response;
+            if (!status.is2xxOK()){
+                throw k2::dto::PlogCreateError("unable to create plog for WAL");
+            }
+            _walPlogPool.push_back(std::move(plogId));
+        }
+        String plogId = _walPlogPool[_walPlogPool.size()-1];
+        _walPlogPool.pop_back();
+        _walInfo = std::move(std::make_pair(plogId, 0));
+
+        // write the metadata of the first WAL plog to metadata log streams
+        Payload temp_payload(Payload::DefaultAllocator);
+        temp_payload.write(std::move(plogId));
+        return _write(std::move(temp_payload), 0);
+    });
+}
+
+seastar::future<std::vector<Payload> >
+LogStream::read(String logStreamName){
+    // read the metadata of metadata log stream from CPO
+    return _cpo.GetMetadataStreamLog(Deadline<>(_cpo_timeout()), logStreamName)
+    .then([&] (auto&& response){
+        auto& [status, resp] = response;
+        if (!status.is2xxOK()) {
+            throw k2::dto::LogStreamMetadataObtainError("unable to obtain metadata log stream");
+        }
+        std::vector<dto::MetadataElement> streamLog = std::move(resp.streamLog);
+        return seastar::do_with(std::move(streamLog), [this] (auto& streamLog){
+            return _client.info(streamLog[streamLog.size()-1].name)
+            .then([&] (auto&& response){
+                auto& [status, resp] = response;
+                if (!status.is2xxOK()) {
+                    throw k2::dto::PlogInfoError("unable to obtain the information of a plog");
+                }
+
+                // read the metadata information of WAL plogs from metadata log stream
+                streamLog[streamLog.size()-1].offset = std::get<0>(resp);
+                std::vector<seastar::future<std::tuple<Status, Payload> > > readFutures;
+                for (auto& log_entry: streamLog){
+                    readFutures.push_back(_client.read(log_entry.name, 0, log_entry.offset));
+                }
+                return seastar::when_all_succeed(readFutures.begin(), readFutures.end());
+            })
+            .then([this] (std::vector<std::tuple<Status, Payload> >&& responses){
+                std::vector<Payload> payloads;
+                for (auto& response: responses){
+                    auto& [status, resp] = response;
+                    if (!status.is2xxOK()){
+                        throw k2::dto::PlogReadError("unable to read a plog");
+                    }
+                    payloads.push_back(std::move(resp));
+                }
+                // read the contents from WAL 
+                return _readContent(std::move(payloads));
+            });
+        });
+        
+    });
+}
+
+seastar::future<std::vector<Payload> >
+LogStream::_readContent(std::vector<Payload> payloads){
+    std::vector<dto::MetadataElement> streamLog;
+
+    //Format: plogId1, offset1, plogId2, offset2... 

```

Comment:
if this is the expected format for elements in the log, you should create a struct for these elements and read/write from the payload in terms of structs, not individual fields. In fact i think your MetadataElement is exactly that